parlai/mturk/core/mturk_data_handler.py

```diff
@@ -0,0 +1,474 @@
+        id string PRIMARY KEY,
+        created integer NOT NULL,
+        maximum integer NOT NULL,
+        completed integer NOT NULL,
+        failed integer NOT NULL
+    );
+    """)
+
+# Worker data table:
+CREATE_WORKER_DATA_SQL_TABLE = (
+    """CREATE TABLE IF NOT EXISTS workers (
+        id string PRIMARY KEY,
+        accepted integer NOT NULL,
+        disconnected integer NOT NULL,
+        completed integer NOT NULL,
+        approved integer NOT NULL,
+        rejected integer NOT NULL
+    );
+    """)
+
+# HIT data table:
+CREATE_HIT_DATA_SQL_TABLE = (
+    """CREATE TABLE IF NOT EXISTS hits (
+        id string PRIMARY KEY,
+        expiration integer NOT NULL,
+        hit_status string,
+        assignments_pending int,
+        assignments_available int,
+        assignments_complete int,
+        run_id string,
+        FOREIGN KEY (run_id) REFERENCES runs (id)
+    );
+    """)
+
+# Assignment data table: (as one HIT can technically have multiple assignments)
+CREATE_ASSIGN_DATA_SQL_TABLE = (
+    """CREATE TABLE IF NOT EXISTS assignments (
+        id string PRIMARY KEY,
+        status string,
+        approve_time int,
+        worker_id string,
+        hit_id string,
+        FOREIGN KEY (worker_id) REFERENCES workers (id),
+        FOREIGN KEY (hit_id) REFERENCES hits (id)
+    );
+    """)
+
+# pairing data table: (reflects one worker<->assignment pairing)
+CREATE_PAIRING_DATA_SQL_TABLE = (
+    """CREATE TABLE IF NOT EXISTS pairings (
+        status string,
+        onboarding_start int,
+        onboarding_end int,
+        task_start int,
+        task_end int,
+        conversation_id string,
+        bonus_amount int,
+        bonus_text string,
+        bonus_paid boolean,
+        worker_id string,
+        assignment_id string,
+        run_id string,
+        FOREIGN KEY (worker_id) REFERENCES workers (id),
+        FOREIGN KEY (assignment_id) REFERENCES assignments (id),
+        FOREIGN KEY (run_id) REFERENCES runs (id)
+    );
+    """)
+
+
+class MTurkDataHandler():
+    """Handles logging data to and reading data from a SQLite3 table for
+    observation across processes and for controlled restarts
+    """
+    def __init__(self, task_group_id=None, file_name='pmt_data.db'):
+        self.db_path = os.path.join(parent_dir, file_name)
+        self.conn = None
+        self.task_group_id = task_group_id
+        self.create_default_tables()
+
+    def _get_connection(self):
+        if self.conn is None:
+            try:
+                conn = sqlite3.connect(self.db_path)
+                self.conn = conn
+            except sqlite3.Error as e:
+                shared_utils.print_and_log(
+                    logging.ERROR,
+                    "Could not get db connection, failing: {}".format(repr(e)),
+                    should_print=True)
+                raise e
+        return self.conn
+
+    def _format_pairing(self, pairing_result):
+        return {
```

Comment:
if you wanted to be fancy, you could do `dict(zip(['status', ..., 'run_id'], pairing_result))`